feat: XAR installation from gulp source-stream

line-o opened the pull request:
closes #126
closes #124 
closes #37 

BREAKING CHANGE: The next version of gulp-exist will no longer be fully compatible with exist versions 2.2 and 3.6.1

- createClient returns new method `install`
    - If provided with the unique package identifier of a xar it allows
      to install a XAR package in the connected existdb.
      The contents and filename are read from a VinylFileStream (gulp.src).
- add type definitions to all user facing functions.
  This will give fair autocompletion and
  hints in  IDEs that read this information (eg. VSCode).

- reorganize index.js
  - move and group exports at the end
  - bind gulp-exist methods to XMLRPCClient. This  gets rid
    of one wrapping function each.

- update to @existdb/node-exist 3.0.1

- add small package test-app.xar for testing
- add tests for package installation

- adapt example gulpfile.js
  - require local gulp-exist to see hints in action
  - use install method in example `installXar` task
- add install section with example to README.MD

### Example Installation

```js
 const { src } = require('gulp')
 const { createClient } = require('@existdb/gulp-exist')

 // override defaults
 const exist = createClient({
     basic_auth: {
         user: 'admin',
         pass: ''
     }
 })
 // this MUST be the unique package identifier of the XAR you want to install 
 const packageUri = 'http://exist-db.org/apps/test-app'

 function install () {
     return src('spec/files/test-app.xar')
         .pipe(exist.install({ packageUri }))
 }
 ```

line-o committed:
feat: XAR installation from gulp source-stream

- createClient returns new method `install`
    - If provided with the unique package identifier of a xar it allows
      to install a XAR package in the connected existdb.
      The contents and filename are read from a VinylFileStream (gulp.src).
- add type definitions to all user facing functions.
  This will give fair autocompletion and
  hints in  IDEs that read this information (eg. VSCode).

- reorganize index.js
  - move and group exports at the end
  - bind gulp-exist methods to XMLRPCClient. This  gets rid
    of one wrapping function each.

- update to @existdb/node-exist 3.0.1

- add small package test-app.xar for testing
- add tests for package installation

- adapt example gulpfile.js
  - require local gulp-exist to see hints in action
  - use install method in example `installXar` task
- add install section with example to README.MD

line-o committed:
fix(ci): do not test on unsupported versions

BREAKING CHANGE: exist versions 2.2 and 3.6.1 are no longer officially supported.

line-o commented:
What do you think @duncdrum ?

duncdrum commented on spec/examples/gulpfile.js:
```diff
@@ -1,22 +1,23 @@
 const { src, dest, watch, series, parallel, lastRun } = require('gulp')
-const { createClient } = require('@existdb/gulp-exist')
-const { connect } = require('@existdb/node-exist')
+const { createClient } = require('../../index')
```
Should this not be `require(‘@existdb/gulp-exist’)` in the example file, as it’s what users are likely to do?

line-o replied:
Yeah, you are probably right. I changed it to be able to see the code hints.

duncdrum reviewed (commented):
Two instances where I thought Switches would be more legible but not a showstopper. One question about the example file. Otherwise looks great.

line-o commented:
@duncdrum where should we switch to `switch`?

line-o committed:
cleanup(example): require the proper package

line-o committed:
doc(readme): add prerequisites and installation

closes #37

- add prerequisites section
- add installation section
- improve usage section
- add API section headline
- fix formatting

olvidalo commented:
:tada: This PR is included in version 4.0.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@existdb/gulp-exist/v/4.0.0)
- [GitHub release](https://github.com/eXist-db/gulp-exist/releases/tag/v4.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
